File: fsharp/Monkey/Lexer.fs
module Monkey.Lexer

open System

type Token =
    | Ident of string
    | Int of int
    | Illegal
    | Eof
    | Equal
    | Plus
    | Comma
    | Semicolon
    | LParen
    | RParen
    | LSquirly
    | RSquirly
    | Function
    | Let

type Code =
    { Source: string
      Position: int
      ReadPosition: int
      Char: char }


let isLetter char =
    char >= 'a' && char <= 'z'
    || char >= 'A' && char <= 'Z'
    || char = '_'

let isNumber char = char >= '0' && char <= '9'

let readChar code =
    let newChar =
        if code.ReadPosition >= code.Source.Length then
            Char.MinValue
        else
            code.Source[code.ReadPosition]

    { code with
        Position = code.ReadPosition
        ReadPosition = code.ReadPosition + 1
        Char = newChar }
        
let toCode source =
    let code =
        { Source = source
          Position = 0
          ReadPosition = 0
          Char = Char.MinValue }

    code |> readChar

let rec skipWhitespace code =
    if code.Char = ' '
        || code.Char = '\t'
        || code.Char = '\n'
        || code.Char = '\r' then
        readChar code |> skipWhitespace
    else
        code

let readWhile code predicate =
    let start = code.Position

    let rec read code =
        if code.Char |> predicate then
            readChar code |> read
        else
            code

    let code = read code
    code, code.Source.Substring(start, code.Position - start)

let readIdentifier code =
    let code, literal = readWhile code isLetter

    let ident =
        match literal with
        | "fn" -> Function
        | "let" -> Let
        | _ -> Ident(literal)
    code, ident

let readNumber code =
    let code, literal = readWhile code isNumber
    code, Int(Int32.Parse(literal))

let nextToken code =
    let code = code |> skipWhitespace

    let advance result = code |> readChar, result

    match code.Char with
    | Char.MinValue -> (code, Eof)
    | '=' -> advance Equal
    | '+' -> advance Plus
    | ',' -> advance Comma
    | ';' -> advance Semicolon
    | '(' -> advance LParen
    | ')' -> advance RParen
    | '{' -> advance LSquirly
    | '}' -> advance RSquirly
    | char when char |> isLetter -> code |> readIdentifier
    | char when char |> isNumber -> code |> readNumber
    | _ -> advance Illegal


let tokenizeCode code =
    let rec loop code =
        seq {
            match code |> nextToken with
            | _, Eof -> yield Eof
            | code, token -> 
                yield token
                yield! loop code
        }
    loop code



let tokenize source = source |> toCode |> tokenizeCode

File: csharp/Monkey/Lexer.cs
namespace Monkey;

public static class Lexer
{
    public static IEnumerable<Token> ParseTokens(string input)
    {
        var current = 0;
        
        while (Next(input.AsSpan(current)) is (token: var token and not EoF, length: var length))
        {
            current += length;
            yield return token;
        }

        yield return Token.EoF;
    }
    
    private static (Token token, int length) Next(ReadOnlySpan<char> input)
    {
        var whiteSpaceLength = ReadWhile(input, char.IsWhiteSpace);
        var (token, length) = MatchToken(input[whiteSpaceLength..]);
        return (token, length + whiteSpaceLength);
    }
    
    private static (Token token, int length) MatchToken(ReadOnlySpan<char> input)
    {
        return input switch
        {                
            ['!', '=', ..] => (Token.NotEqual, 2),
            ['=', '=', ..] => (Token.Equal, 2),
            ['=', ..] => (Token.Assign, 1),
            ['+', ..] => (Token.Plus, 1),
            ['-', ..] => (Token.Dash, 1),
            ['!', ..] => (Token.Bang, 1),
            ['*', ..] => (Token.Asterisk, 1),
            ['/', ..] => (Token.ForwardSlash, 1),
            ['<', ..] => (Token.LessThan, 1),
            ['>', ..] => (Token.GreaterThan, 1),
            [',', ..] => (Token.Comma, 1),
            [';', ..] => (Token.Semicolon, 1),
            ['(', ..] => (Token.LParen, 1),
            [')', ..] => (Token.RParen, 1),
            ['{', ..] => (Token.LSquirly, 1),
            ['}', ..] => (Token.RSquirly, 1),
            
            [var c, ..]
                when char.IsLetter(c)
                && ReadWhile(input, char.IsLetterOrDigit) is >0 and var length 
                => (MatchKeywordOrIdentifier(input[..length]), length),
            
            _ when ReadWhile(input, char.IsDigit) is >0 and var length 
                => (Token.Integer(int.Parse(input[..length])), length),
            
            [var c, ..] => (Token.Illegal(c), 1),
            
            [] => (Token.EoF, 0)
        };
    }

    private static Token MatchKeywordOrIdentifier(ReadOnlySpan<char> input)
    {
        return input switch
        {
            "fn" => Token.Function,
            "let" => Token.Let,
            "true" => Token.True,
            "false" => Token.False,
            "if" => Token.If,
            "else" => Token.Else,
            "return" => Token.Return,
            _ => Token.Identifier(input.ToString())
        };
    }
    
    private static int ReadWhile(ReadOnlySpan<char> input, Func<char, bool> predicate)
    {
        var i = 0;
        
        while (i < input.Length && predicate(input[i]))
        {
            i++;
        }

        return i;
    }
}